crates/nostr/src/message/client.rs

```diff
@@ -26,6 +28,29 @@ pub enum ClientMessage {
     Auth(Box<Event>),
 }
 
+impl Serialize for ClientMessage {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let json_string = self.as_json();
+        let json_value: Value = serde_json::from_str(&json_string).unwrap();
+        json_value.serialize(serializer)
+    }
+}
+
+impl<'de> Deserialize<'de> for ClientMessage {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let json_value = Value::deserialize(deserializer)?;
+        let json_string = serde_json::to_string(&json_value).unwrap();
```

Comment:
Replace with:

```rust
let json_value = Value::deserialize(deserializer)?;
ClientMessage::from_value(json_value).map_err(Error::custom)
```

Add `from_value` method:

```rust
fn from_value(msg: Value) -> Result<Self, MessageHandleError> {
   let v = msg
        .as_array()
        .ok_or(MessageHandleError::InvalidMessageFormat)?;

    if v.is_empty() {
        return Err(MessageHandleError::InvalidMessageFormat);
    }

    let v_len: usize = v.len();

    // Event
    // ["EVENT", <event JSON>]
    if v[0] == "EVENT" {
        if v_len != 2 {
            return Err(MessageHandleError::InvalidMessageFormat);
        }
        let event = Event::from_json(v[1].to_string())
            .map_err(|_| MessageHandleError::JsonDeserializationFailed)?;
        return Ok(Self::new_event(event));
    }

    // Req
    // ["REQ", <subscription_id>, <filter JSON>, <filter JSON>...]
    if v[0] == "REQ" {
        if v_len == 2 {
            let subscription_id: SubscriptionId = serde_json::from_value(v[1].clone())
                .map_err(|_| MessageHandleError::JsonDeserializationFailed)?;
            return Ok(Self::new_req(subscription_id, Vec::new()));
        } else if v_len == 3 {
            let subscription_id: SubscriptionId = serde_json::from_value(v[1].clone())
                .map_err(|_| MessageHandleError::JsonDeserializationFailed)?;
            let filters: Vec<Filter> = serde_json::from_value(Value::Array(v[2..].to_vec()))
                .map_err(|_| MessageHandleError::JsonDeserializationFailed)?;
            return Ok(Self::new_req(subscription_id, filters));
        } else {
            return Err(MessageHandleError::InvalidMessageFormat);
        }
    }

    // Close
    // ["CLOSE", <subscription_id>]
    if v[0] == "CLOSE" {
        if v_len != 2 {
            return Err(MessageHandleError::InvalidMessageFormat);
        }

        let subscription_id: SubscriptionId = serde_json::from_value(v[1].clone())
            .map_err(|_| MessageHandleError::JsonDeserializationFailed)?;

        return Ok(Self::close(subscription_id));
    }

    // Auth
    // ["AUTH", <event JSON>]
    if v[0] == "AUTH" {
        if v_len != 2 {
            return Err(MessageHandleError::InvalidMessageFormat);
        }
        let event = Event::from_json(v[1].to_string())
            .map_err(|_| MessageHandleError::JsonDeserializationFailed)?;
        return Ok(Self::new_auth(event));
    }

    Err(MessageHandleError::InvalidMessageFormat)
}
```

And update `from_json` method:

```rust
pub fn from_json<S>(msg: S) -> Result<Self, MessageHandleError>
where
    S: Into<String>,
{
    let msg: &str = &msg.into();

    log::trace!("{}", msg);

    let value: Value =
        serde_json::from_str(msg).map_err(|_| MessageHandleError::JsonDeserializationFailed)?;

    Self::from_value(value)
}
```